src/coreclr/jit/valuenum.cpp

```diff
@@ -10956,6 +11033,61 @@ void Compiler::fgValueNumberAddExceptionSet(GenTree* tree)
 }
 
 #ifdef DEBUG
+//------------------------------------------------------------------------
+// fgDebugCheckExceptionSets: Verify the exception sets on trees.
+//
+// This function checks that the node's exception set is a superset of
+// the exception sets of its operands.
+//
+void Compiler::fgDebugCheckExceptionSets()
+{
+    struct ExceptionSetsChecker
+    {
+        static void CheckTree(GenTree* tree, ValueNumStore* vnStore)
+        {
+            // We will fail to VN some PHI_ARGs - their values may not
+            // be known at the point we number them because of loops.
+            assert(tree->gtVNPair.BothDefined() || tree->OperIs(GT_PHI_ARG));
+
+            ValueNumPair operandsExcSet = vnStore->VNPForEmptyExcSet();
+            tree->VisitOperands([&](GenTree* operand) -> GenTree::VisitResult {
+
+                CheckTree(operand, vnStore);
+
+                ValueNumPair operandVNP = operand->gtVNPair.BothDefined() ? operand->gtVNPair : vnStore->VNPForVoid();
+                operandsExcSet          = vnStore->VNPUnionExcSet(operandVNP, operandsExcSet);
+
+                return GenTree::VisitResult::Continue;
+            });
+
+            // Currently, we fail to properly maintain the exception sets for trees with user
+            // calls or assignments.
+            if ((tree->gtFlags & (GTF_ASG | GTF_CALL)) != 0)
+            {
+                return;
+            }
+
+            ValueNumPair nodeExcSet = vnStore->VNPExceptionSet(tree->gtVNPair);
+            assert(vnStore->VNExcIsSubset(nodeExcSet.GetLiberal(), operandsExcSet.GetLiberal()));
+            assert(vnStore->VNExcIsSubset(nodeExcSet.GetConservative(), operandsExcSet.GetConservative()));
+        }
```

Comment:
I'm curious, is there a reason you opted for this manually recursive approach instead of using `GenTreeVisitor`?

Reply:
Not really; the `VisitOperands` + recursion approach was just "obvious".

I suppose the way to do the same with the visitor would be to check that the node's exception set is a subset of its `user`.

I know there is some desire to in some way make the visitor itself non-recursive, but it is not clear to me if that's something feasible.

Reply:
> I suppose the way to do the same with the visitor would be to check that the node's exception set is a subset of its user.

Right, that might be the simplest way (or you could keep the flags as a member variable and check them in post-order).